src/rgw/rgw_ioc_dispatch.cc

```diff
@@ -0,0 +1,270 @@
+#endif /* O_BINARY */
+
+namespace ioc {
+
+namespace detail {
+/// unique_ptr with custom deleter for struct aiocb
+struct AioCBDeleter {
+  void operator()(struct aiocb* c) {
+        if(c->aio_fildes > 0) ::close(c->aio_fildes);
+        delete c;
+  }
+};
+
+using unique_aio_cb_ptr = std::unique_ptr<struct aiocb, AioCBDeleter>;
+
+template <typename Result>
+struct AsyncFileReadOp {
+  bufferlist result;
+  unique_aio_cb_ptr aio_cb;
+  using Signature = void(boost::system::error_code, Result);
+  using Completion = ceph::async::Completion<Signature, AsyncFileReadOp<Result>>;
+  int init(std::string file_path, off_t read_ofs, off_t read_len, void* arg) {
+    aio_cb.reset(new struct aiocb);
+    memset(aio_cb.get(), 0, sizeof(struct aiocb));
+    aio_cb->aio_fildes = TEMP_FAILURE_RETRY(::open(file_path.c_str(), O_RDONLY|O_CLOEXEC|O_BINARY));
+    if(aio_cb->aio_fildes < 0) {
+      int err = errno;
+      dout(2) << "IOC:AsyncFileReadOp can't open " << file_path << ": " << cpp_strerror(err) << dendl;
+      return -err;
+    }
+
+    bufferptr bp(read_len);
+    aio_cb->aio_buf = bp.c_str();
+    result.append(std::move(bp));
+
+    aio_cb->aio_nbytes = read_len;
+    aio_cb->aio_offset = read_ofs;
+    aio_cb->aio_sigevent.sigev_notify = SIGEV_THREAD;
+    aio_cb->aio_sigevent.sigev_notify_function = aio_dispatch;
+    aio_cb->aio_sigevent.sigev_notify_attributes = nullptr;
+    aio_cb->aio_sigevent.sigev_value.sival_ptr = arg;
+
+    return 0;
+  }
+
+  static void aio_dispatch(sigval_t sigval) {
+    auto p = std::unique_ptr<Completion>{static_cast<Completion*>(sigval.sival_ptr)};
+    auto op = std::move(p->user_data);
+    const int ret = -aio_error(op.aio_cb.get());
+    boost::system::error_code ec;
+    if (ret < 0) {
+        ec.assign(-ret, boost::system::system_category());
+    }
+
+    ceph::async::dispatch(std::move(p), ec, std::move(op.result));
+  }
+
+  template <typename Executor1, typename CompletionHandler>
+  static auto create(const Executor1& ex1, CompletionHandler&& handler) {
+    auto p = Completion::create(ex1, std::move(handler));
+    return p;
+  }
+};
+} // namespace detail
+
+template <typename ExecutionContext, typename CompletionToken>
+auto async_read(ExecutionContext& ctx, const std::string& file_path,
+                off_t read_ofs, off_t read_len, CompletionToken&& token) {
+  using Op = detail::AsyncFileReadOp<bufferlist>;
+  using Signature = typename Op::Signature;
+  boost::asio::async_completion<CompletionToken, Signature> init(token);
+  auto p = Op::create(ctx.get_executor(), init.completion_handler);
+  auto& op = p->user_data;
+  int ret = op.init(file_path, read_ofs, read_len, p.get());
+
+  if(0 == ret) {
+    ret = aio_read(op.aio_cb.get());
+  }
+
+  if(ret < 0) {
+    auto ec = boost::system::error_code{-ret, boost::system::system_category()};
+    ceph::async::post(std::move(p), ec, bufferlist{});
+  } else {
+    p.release();
+  }
+  return init.result.get();
+}
+
+namespace {
```

Comment:
is this the anonymous namespace?  why?

Reply:
The usage I learned from rgw_aio.cc, it puts the handler and some functions into the anonymous space.  
Removing this anonymous space has no effect on this piece of code. Do I need to remove it?